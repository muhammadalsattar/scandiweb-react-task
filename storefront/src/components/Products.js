import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import Navbar from "./Navbar";
import { addToCart } from "../actions/cart";
import { setActiveProduct, setLoadedProducts } from "../actions/products";
import {pickPrice} from "../utils/utils";

export class Products extends React.Component {
  updateCart = e => {
    e.preventDefault();

    const product = this.props.products.find(product => product.id === e.target.id);
    const selectedAttr = product.attributes.map(attribute => {
      return {
        name: attribute.name,
        value: attribute.items[0].value
      };
    });
    const newProduct = {...product, selectedAttr, quantity: 1};
    this.props.addToCart(newProduct);

    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    });
    document.querySelector(".cart-overlay").removeAttribute("hidden");
    document.querySelector(".page-body").classList.add("faded")
  }

  loadMore = e => {
    e.preventDefault();
    this.props.setLoadedProducts(this.props.loadedProducts, this.props.allProducts.length);
  }
  
  render() {
    return (
      <div className="products-page">
        <Navbar />
        <div className="products page-body">
          <h1 className="category-name">{this.props.defaultCategory}</h1>
          <div className="products-list">
            {this.props.products.map(product =>
              (
              <div className={"product-item " + product.inStock} key={product.id}>
                {product.inStock && <h2 className="out-stock">out of stock</h2>}
                <div className="product-image">
                  <Link to={`/product?id=${product.id}`} ><img src={product.gallery[0]} alt={product.name}></img></Link>
                </div>
                <div className="product-info">
                  <p className="product-name">{product.name}</p>
                  <p className="product-price">{this.props.defaultCurrency.symbol}{pickPrice(product.prices, this.props.defaultCurrency)}</p>
                </div>
                <button className="bag-button"><img id={product.id} onClick={this.updateCart} src={process.env.PUBLIC_URL+"/circle-icon.png"} alt="cart-circle"></img></button>
              </div>
              )
            )}
          </div>
        </div>
        <button className="load-more" onClick={this.loadMore}>Load More</button>
    </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addToCart: product => {
      dispatch(addToCart(product));
    },
    setActiveProduct: productID => {
      dispatch(setActiveProduct(productID));
    },
    setLoadedProducts: (loadedProducts, productsLength) => {
      if(loadedProducts + 5 > productsLength) {
        dispatch(setLoadedProducts(productsLength));
      } else {
        dispatch(setLoadedProducts(loadedProducts + 5));
      }
    }
  };
};

const mapStateToProps = state => {
  return {
    products: state.categories.defaultCategory.name === "all"?
    state.products.products.slice(0, state.products.loadedProducts).map(product => {
      return product;
    }):
    state.products.products.slice(0, state.products.loadedProducts).filter(product=>product.category === state.categories.defaultCategory.name).map(product => {
      return product;
    }),
    allProducts: state.products.products,
    loadedProducts: state.products.loadedProducts,
    defaultCategory: state.categories.defaultCategory.name,
    cart: state.cart.cart,
    defaultCurrency: state.currencies.defaultCurrency
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Products);