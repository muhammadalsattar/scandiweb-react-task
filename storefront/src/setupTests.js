const currencies = [
    {
        label: 'USD',
        symbol: '$',
    },
    {
        label: 'GBP',
        symbol: 'Â£',
    },
];
const categories = [
    {
        name: 'category1',
    },
    {
        name: 'category2',
    },
];
const attributes = [
    {
        id: 'Size',
        name: 'Size',
        type: 'text',
        items: [
            {
                id: 40,
                value: '40',
                displayValue: '40',
            },
        ],
    },
    {
        id: 'Color',
        name: 'Color',
        type: 'swatch',
        items: [
            {
                id: 'red',
                value: 'red',
                displayValue: 'Red',
            },
        ],
    },
];
const gallery = [
    'https://dummyimage.com/600x400/000/fff',
    'https://dummyimage.com/600x400/000/ggg',
    'https://dummyimage.com/600x400/000/hhh',
];
const prices = [
    {
        amount: 518.47,
        currency: currencies[0],
    },
    {
        amount: 372.67,
        currency: currencies[1],
    },
];
const products = [
    {
        id: '1',
        name: 'product1',
        brand: 'brand1',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '40' }],
    },
    {
        id: '2',
        name: 'product2',
        brand: 'brand2',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'red' }],
    },
    {
        id: '3',
        name: 'product3',
        brand: 'brand3',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '41' }],
    },
    {
        id: '4',
        name: 'product4',
        brand: 'brand4',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'orange' }],
    },
    {
        id: '5',
        name: 'product5',
        brand: 'brand5',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[0],
        inStock: true,
        quantity: 1,
        selectedAttr: [{ name: 'Size', value: '42' }],
    },
    {
        id: '6',
        name: 'product6',
        brand: 'brand6',
        attributes,
        description: '<p>dummy text</p>',
        gallery,
        prices,
        category: categories[1],
        inStock: false,
        quantity: 1,
        selectedAttr: [{ name: 'Color', value: 'green' }],
    },
];

export const initialState = {
    currencies,
    categories,
    products,
};
