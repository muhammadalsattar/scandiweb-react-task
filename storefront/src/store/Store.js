import { configureStore, combineReducers } from '@reduxjs/toolkit';
import cartReducer from '../reducers/cartReducer';
import productsReducer from '../reducers/productsReducer';
import categoriesReducer from '../reducers/categoriesReducer';
import currencyReducer from '../reducers/currencyReducer';

const rootReducer = combineReducers({
    cart: cartReducer,
    products: productsReducer,
    categories: categoriesReducer,
    currencies: currencyReducer,
});
// creating store and combining reducers
const Store = configureStore({
    reducer: rootReducer,
});

export default Store;
