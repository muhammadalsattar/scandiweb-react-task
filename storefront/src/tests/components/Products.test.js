import React from "react";
import {render, screen} from "@testing-library/react";
import { Provider } from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import Store from "../../store/Store.js";
import {Products} from "../../components/Products.js";
import { initialState } from "../../setupTests.js"

test("Products renders correctly", () => {
    const { undefined, container } = render(
        <Provider store={Store}>
        <Router><Products cart={initialState.cart} defaultCurrency={initialState.currencies[0]} products={initialState.products} defaultCategory={initialState.categories[0].name}/></Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getAllByText(/out of stock/i)).toHaveLength(5);
});