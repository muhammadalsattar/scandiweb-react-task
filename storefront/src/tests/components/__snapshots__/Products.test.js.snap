// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products renders correctly 1`] = `
<div>
  <div
    class="products-page"
  >
    <nav
      class="navbar"
    >
      <div
        class="categories"
      />
      <div
        class="brand-logo"
      >
        <a
          href="/"
        >
          <img
            alt="logo-icon"
            src="/logo.png"
          />
        </a>
      </div>
      <div
        class="dropdowns"
      >
        <div
          class="currency-div"
        >
          <button
            class="currency-icon"
          >
            <span
              class="currency-drop-icon"
            >
              âˆ§
            </span>
          </button>
          <ul
            class="currency-dropdown"
            hidden=""
          />
        </div>
        <div
          class="cart-div"
        >
          <button
            class="cart-img"
          >
            <img
              alt="cart-img"
              src="/cart.png"
            />
          </button>
          <button
            class="cart-count"
          >
            0
          </button>
          <div
            class="cart-overlay"
            hidden=""
          >
            <p>
              <b>
                My Bag, 
              </b>
              0
               items
            </p>
            <div
              class="cart-total"
            >
              <p>
                Total
              </p>
              <p>
                0.00
              </p>
            </div>
            <div
              class="bag-checkout"
            >
              <div
                class="view-bag"
              >
                <a
                  href="/cart"
                >
                  <button>
                    View Bag
                  </button>
                </a>
              </div>
              <div
                class="checkout"
              >
                <button>
                  Check out
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="products page-body"
    >
      <h1
        class="category-name"
      >
        all
      </h1>
      <div
        class="products-list"
      >
        <div
          class="product-item true"
        >
          <h2
            class="out-stock"
          >
            out of stock
          </h2>
          <div
            class="product-image"
          >
            <a
              href="/product?id=huarache-x-stussy-le"
            >
              <img
                alt="Nike Air Huarache Le"
                src="https://cdn.shopify.com/s/files/1/0087/6193/3920/products/DD1381200_DEOA_2_720x.jpg?v=1612816087"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              Nike Air Huarache Le
            </p>
            <p
              class="product-price"
            >
              $
              144.69
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="huarache-x-stussy-le"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item true"
        >
          <h2
            class="out-stock"
          >
            out of stock
          </h2>
          <div
            class="product-image"
          >
            <a
              href="/product?id=jacket-canada-goosee"
            >
              <img
                alt="Jacket"
                src="https://images.canadagoose.com/image/upload/w_480,c_scale,f_auto,q_auto:best/v1576016105/product-image/2409L_61.jpg"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              Jacket
            </p>
            <p
              class="product-price"
            >
              $
              518.47
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="jacket-canada-goosee"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item false"
        >
          <div
            class="product-image"
          >
            <a
              href="/product?id=ps-5"
            >
              <img
                alt="PlayStation 5"
                src="https://images-na.ssl-images-amazon.com/images/I/510VSJ9mWDL._SL1262_.jpg"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              PlayStation 5
            </p>
            <p
              class="product-price"
            >
              $
              844.02
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="ps-5"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item false"
        >
          <div
            class="product-image"
          >
            <a
              href="/product?id=xbox-series-s"
            >
              <img
                alt="Xbox Series S 512GB"
                src="https://images-na.ssl-images-amazon.com/images/I/71vPCX0bS-L._SL1500_.jpg"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              Xbox Series S 512GB
            </p>
            <p
              class="product-price"
            >
              $
              333.99
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="xbox-series-s"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item true"
        >
          <h2
            class="out-stock"
          >
            out of stock
          </h2>
          <div
            class="product-image"
          >
            <a
              href="/product?id=apple-imac-2021"
            >
              <img
                alt="iMac 2021"
                src="https://store.storeimages.cdn-apple.com/4982/as-images.apple.com/is/imac-24-blue-selection-hero-202104?wid=904&hei=840&fmt=jpeg&qlt=80&.v=1617492405000"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              iMac 2021
            </p>
            <p
              class="product-price"
            >
              $
              1688.03
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="apple-imac-2021"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item true"
        >
          <h2
            class="out-stock"
          >
            out of stock
          </h2>
          <div
            class="product-image"
          >
            <a
              href="/product?id=apple-iphone-12-pro"
            >
              <img
                alt="iPhone 12 Pro"
                src="https://store.storeimages.cdn-apple.com/4982/as-images.apple.com/is/iphone-12-pro-family-hero?wid=940&amp;hei=1112&amp;fmt=jpeg&amp;qlt=80&amp;.v=1604021663000"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              iPhone 12 Pro
            </p>
            <p
              class="product-price"
            >
              $
              1000.76
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="apple-iphone-12-pro"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item false"
        >
          <div
            class="product-image"
          >
            <a
              href="/product?id=apple-airpods-pro"
            >
              <img
                alt="AirPods Pro"
                src="https://store.storeimages.cdn-apple.com/4982/as-images.apple.com/is/MWP22?wid=572&hei=572&fmt=jpeg&qlt=95&.v=1591634795000"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              AirPods Pro
            </p>
            <p
              class="product-price"
            >
              $
              300.23
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="apple-airpods-pro"
              src="/circle-icon.png"
            />
          </button>
        </div>
        <div
          class="product-item true"
        >
          <h2
            class="out-stock"
          >
            out of stock
          </h2>
          <div
            class="product-image"
          >
            <a
              href="/product?id=apple-airtag"
            >
              <img
                alt="AirTag"
                src="https://store.storeimages.cdn-apple.com/4982/as-images.apple.com/is/airtag-double-select-202104?wid=445&hei=370&fmt=jpeg&qlt=95&.v=1617761672000"
              />
            </a>
          </div>
          <div
            class="product-info"
          >
            <p
              class="product-name"
            >
              AirTag
            </p>
            <p
              class="product-price"
            >
              $
              120.57
            </p>
          </div>
          <button
            class="bag-button"
          >
            <img
              alt="cart-circle"
              id="apple-airtag"
              src="/circle-icon.png"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
